Reference to quoted local member fails to in quote pickling

## Compiler version

3.3.1-RC1-bin-20230313-f28d708-NIGHTLY and before


## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
import scala.quoted.*

def test(using Quotes): Expr[Unit] =
  '{
    trait C:
      val d: D
    trait D
    val c: C = ???
    ${
      val expr = '{
        val cRef: c.type = ???
        cRef.d
        ()
      }
      expr
    }
  }
```

## Output (click arrow to expand)
<details>

```scala
Exception in thread "main" 
  unhandled exception while running pickleQuotes on tests/neg-macros/path-dependent-type-capture.scala

  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/lampepfl/dotty/issues/new/choose

     while compiling: <no file>
        during phase: <no phase>
                mode: Mode(ImplicitsEnabled)
     library version: version 2.13.10
    compiler version: version 3.3.1-RC1-bin-SNAPSHOT-nonbootstrapped-git-29e229b
            settings: -Vprint List(splicing, staging) -classpath /Users/nicolasstucki/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.10/scala-library-2.13.10.jar:/Users/nicolasstucki/Github/dotty/library/../out/bootstrap/scala3-library-bootstrapped/scala-3.3.1-RC1-bin-SNAPSHOT-nonbootstrapped/scala3-library_3-3.3.1-RC1-bin-SNAPSHOT.jar -d t/out -pagewidth 120

                tree: EmptyTree
       tree position: :<unknown>
           tree type: <notype>
              symbol: val <none>
           call site: package <root> in module class <root>

  == Source file context for tree position ==
java.lang.AssertionError: assertion failed: unresolved symbols: value cRef (line 11) #15683 when pickling tests/neg-macros/path-dependent-type-capture.scala
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
        at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:809)
        at dotty.tools.dotc.quoted.PickledQuotes$.pickle(PickledQuotes.scala:222)
        at dotty.tools.dotc.quoted.PickledQuotes$.pickleQuote(PickledQuotes.scala:34)
        at dotty.tools.dotc.transform.PickleQuotes$.pickleAsTasty$1(PickleQuotes.scala:291)
        ...
```
</details>

--- a/compiler/src/dotty/tools/dotc/transform/Splicing.scala
+++ b/compiler/src/dotty/tools/dotc/transform/Splicing.scala
@@ -208,6 +208,14 @@ class Splicing extends MacroTransform:
 
     override def transform(tree: tpd.Tree)(using Context): tpd.Tree =
       tree match
+        case tree: Select if tree.isTerm && isCaptured(tree.symbol) =>
+          tree.symbol.allOverriddenSymbols.find(sym => !isCaptured(sym.owner)) match
+            case Some(sym) =>
+              // virtualize call on overridden symbol that is not defined in a non static class
+              transform(tree.qualifier.select(sym))
+            case _ =>
+              report.error(em"Can not use reference to staged local ${tree.symbol} defined in an outer quote.\n\nThis can work if ${tree.symbol.owner} would extend a top level interface that defines ${tree.symbol}.", tree)
+              tree
         case tree: RefTree =>
           if tree.isTerm then
             if isCaptured(tree.symbol) then

--- a/tests/neg-macros/i17103.scala
+++ b/tests/neg-macros/i17103.scala
@@ -1,0 +1,16 @@
+import scala.quoted.*
+
+def test(using Quotes): Expr[Unit] =
+  '{
+    trait C:
+      def d: Int
+    val c: C = ???
+    ${
+      val expr = '{
+        val cRef: c.type = ???
+        cRef.d // error
+        ()
+      }
+      expr
+    }
+  }

--- a/tests/pos-macros/i17103a.scala
+++ b/tests/pos-macros/i17103a.scala
@@ -1,0 +1,21 @@
+import scala.quoted.*
+
+trait C0:
+  def d: Int
+
+def test(using Quotes): Expr[Unit] =
+  '{
+    trait C1 extends C0:
+      def d: Int
+    trait C extends C1:
+      def d: Int
+    val c: C = ???
+    ${
+      val expr = '{
+        val cRef: C = ???
+        cRef.d // calls C0.d
+        ()
+      }
+      expr
+    }
+  }

--- a/tests/pos-macros/i17103b.scala
+++ b/tests/pos-macros/i17103b.scala
@@ -1,0 +1,21 @@
+import scala.quoted.*
+
+trait C0:
+  def d: Int
+
+def test(using Quotes): Expr[Unit] =
+  '{
+    trait C1 extends C0:
+      def d: Int
+    trait C extends C1:
+      def d: Int
+    val c: C = ???
+    ${
+      val expr = '{
+        val cRef: c.type = ???
+        cRef.d // calls C0.d
+        ()
+      }
+      expr
+    }
+  }

--- a/tests/pos-macros/i7405b.scala
+++ b/tests/pos-macros/i7405b.scala
@@ -3,7 +3,7 @@ import scala.quoted.*
 class Foo {
   def f(using Quotes): Expr[Any] = {
     '{
-      trait X {
+      trait X extends A {
         type Y
         def y: Y = ???
       }
@@ -17,3 +17,7 @@ class Foo {
     }
   }
 }
+
+trait A:
+  type Y
+  def y: Y = ???

